Add CES prompts for initial setup and campaign creation

ianlin opened the pull request:
### Changes proposed in this Pull Request:

Closes #882.

This PR implements two CES prompts that will displayed when the user

**Prompt 1**: successfully completes initial setup flow and enters the product feed page for the first time
**Prompt 2**: successfully completes Ads creation flow and clicks "Got it" in the success modal.

### Implementation Details

#### Prompt 1

This can be split into two requirements:

1. Display it when the user enters product feed for the first time after setting up
2. Don't display it unless the success modal is not shown. I.e. dismiss the success modal will display the CES prompt

When the user sees the success modal, there are different paths that they can go:

1. Dismiss the success modal by clicking `X` button, clicking anywhere outside of the modal, or clicking `Maybe later` button.
2. Go to the product feed page in a new tab by clicking `Manage and edit your product feed in WooCommerce`.
3. Go to the page to create paid campaign, and they will be redirected to the dashboard after finish creating campaign.
  a. Eventually the user will go to the product feed page by clicking it, so the behaviour will be just like the point 2 above

In oder to fulfil the requirements above, we need to store a flag `gla-can-onboarding-setup-ces-prompt-open` in the local storage to help us to decide when can we display the CES prompt. The flows will be like this:

**Flow 1** ([Video 1](#video-1))
1. User sees the success modal
2. At the mean time the CES prompt will not be displayed since the success modal is shown
3. The success modal set `gla-can-onboarding-setup-ces-prompt-open` to `true` (actually any value is fine)
4. User dismisses the success modal, then the CES prompt is shown
5. The CES prompt will delete the flag `gla-can-onboarding-setup-ces-prompt-open` from the local storage

**Flow 2** ([Video 2](#video-2))
1. User sees the success modal
2. The success modal set `gla-can-onboarding-setup-ces-prompt-open` to `true`
3. User go to the product feed page in a new tab
4. In the new tab there are no success modal and the flag is in the local storage, the CES prompt is shown
5. The CES prompt deletes  `gla-can-onboarding-setup-ces-prompt-open`

#### Prompt 2

Prompt 2 is simpler, the prompt will only be shown when the user clicks on "Got it" of the campaign creation success modal, so we don't need the flag in local storage.

([Video 3](#video-3))
([Video 4](#video-4))

### Screenshots:

#### <a href="#video-1" id="video-1">Video 1</a> - Dismiss success modal and show the CES prompt

https://user-images.githubusercontent.com/914406/146525605-dcafbc9b-ccf2-4174-964c-27bd67cb35be.mov

#### <a href="#video-2" id="video-2">Video 2</a> - Go to product feed in a new tab

https://user-images.githubusercontent.com/914406/146525728-1862b7e7-9471-45dc-9971-5dfef45498ef.mov

#### <a href="#video-3" id="video-3">Video 3</a> - Click "Got it" on the campaign creation success modal and show the CES

https://user-images.githubusercontent.com/914406/146526005-1f088eb4-61e5-499c-b094-d28b272ba3e6.mov

#### <a href="#video-4" id="video-4">Video 4</a> - Dismiss the campaign creation success modal and the CES won't show

https://user-images.githubusercontent.com/914406/146527229-09e4e458-024a-4a37-90e1-b49ea35d5b28.mov

### Detailed test instructions:

#### Prompt 1, Flow 1

1. Go to the product feed page with the success modal opened: `/wp-admin/admin.php?page=wc-admin&path=/google/product-feed&guide=submission-success`
2. Enable event debugging by typing `localStorage.setItem('debug', 'wc-admin*')` in your web's dev console
3. Dismiss the success modal
4. Check if the CES prompt is shown
5. Check if the events are sent correctly from the dev console

#### Prompt 1, Flow 2

1. ~Follow the first three steps above~ Follow the first **two** steps above (thanks @puntope)
2. Go to the product feed page in a new tab by click `Manage and edit your product feed in WooCommerce` 
3. Check if the CES prompt is shown
4. Check if the events are sent correctly from the dev console
5. Go back to the original tab and dismiss the success modal, the CES prompt should not be shown

#### Prompt 2

1. Go to the GLA dashboard page with the campaign creation modal opened: `/wp-admin/admin.php?page=wc-admin&path=/google/dashboard&guide=campaign-creation-success`
2. Check if the CES prompt is shown if clicking "Got it"
3. Check if the CES prompt is not shown if dismissing the success modal
4. Check if the events are sent correctly from the dev console

<!--
Optional.
Enter a summary of all changes in this Pull Request, which will be added to the changelog if accepted.
Each line should start with `(Fix|Add|Tweak|Update) - `, for example:
> Fix - I took care of something that wasn't working.
> Add - I added something new that's pretty cool.
> Tweak - I made a small change.
> Update - I made big changes to something that wasn't broken.

Leave the "Changelog entry" header in place completely empty, without any summary if no changelog entry is needed.
If you remove the "Changelog entry" header, the title of Pull Request will be used as the changelog entry.
-->
### Changelog entry

> Add - CES prompts for initial setup and campaign creation.

ianlin committed:
Add @woocommerce/customer-effort-score module

ianlin committed:
Move checking qs "guide" from SubmissionSuccessGuide to its parent

Since this logic will also be used for the CES prompt component in
the upcoming implementaion.

ianlin committed:
Use the guide name defined in the constants

ecgan commented:
@ianlin , I converted your PR to draft, since it is still "work in progress". You can create a draft PR without needing to specify "WIP" in the PR title. When the PR is ready for review, just add the reviewers and click on the "Ready for review" button. 🙂

ianlin commented:
@ecgan TIL, thank you! 😄

ianlin committed:
Add a wrapper for window.localStorage

ianlin committed:
Display onboarding setup CES prmopt in the right timing

Requirement:

1. Display it when the user enters product feed for the first time after
	 setting up
2. However, don't display it unless the success modal is not shown. I.e.
	 dismiss the success modal will display the CES prompt

Displaying Logic:

- When the user finishes setting up, they will be in the product feed
	page and the success modal is shown

- There are 2 scenarios:

  - If the user dismiss the success modal, the CES prmopt will be
		displayed

  - If the user clicks "Manage and edit your product feed in
		WooCommerce" and goes to the product feed page in a new tab, the CES
		prompt will be displayed in the new tab

    - If the user dismisses the CES prompt or clicks "Give feedback" on
			the prompt, the CES prompt in the original tab will not be shown
			when the success modal is dismissed

    - But if the user does not dismiss the CES prompt or clicks "Give
			feedback" on the prompt in the new tab, the CES prompt in the
			original tab will be shown when the success modal is dismissed

ianlin committed:
Send CES feedback using event tracking

ianlin committed:
Only display CES prompts when WC tracking is enabled

- DB options: woocommerce_allow_tracking

- wp-admin settings:
	WooCommerce -> Settings -> Advanced -> WooCommerce.com -> Enable tracking

- Frontend checks `window.wcTracks.isEnabled`

ianlin committed:
Replace Ces with CES

Note: CES = Customer Effort Score

ianlin committed:
Move CustomerEffortScorePrompt from product-feed/ to components/

And make it more general so it can be a shared component

ianlin committed:
Display Ads campaign creation CES prompt

This is rather simple. This prompt should appear immediately after user
successfully completes Ads creation flow and clicks "Got it" in the
success modal.

ianlin committed:
Revert "Remove not used `Svgr` mock."

This reverts commit 8ac7159cb6c0afbb373ab65165c2d4f858f16703.

ianlin committed:
Revert "Remove SVG module mapper,"

This reverts commit c101d69bae986b6181604606b9b0611867348bb8.

ianlin committed:
Add tests for js/src/product-feed/index.js

ianlin committed:
Add doc to CustomerEffortScorePrompt component

ianlin committed:
Add tests for js/src/dashboard/campaign-creation-success-guide

puntope commented on js/src/components/customer-effort-score-prompt/index.js:
```diff
@@ -0,0 +1,92 @@
+import { LOCAL_STORAGE_KEYS } from '.~/constants';
+import localStorage from '.~/utils/localStorage';
+
+/**
+ * A CustomerEffortScore wrapper that uses tracks to track the selected
+ * customer effort score.
+ *
+ * @param {Object} props React component props.
+ * @param {string} props.eventContext Context to be used in the CES wrapper events.
+ * @param {string} props.label Text to be displayed in the CES notice and modal.
+ */
```
💅 

Documentation about records is missing

See `js/src/components/contact-information/store-address-card.js`  (as an example)

We are using JSDOC documentation in a way that running `npm run doc:tracking` generates tracking docs automatically.

ianlin replied:
Thanks, added in 2932545 manually as in GLA repo it seems we haven't had `npm run doc:tracking` to generate tracking docs.

puntope replied:
Thanks for pointing out that we don't have that yet here. Didn't see this https://github.com/woocommerce/google-listings-and-ads/issues/117 issue

puntope commented on js/src/dashboard/campaign-creation-success-guide/index.js:
```diff
@@ -34,17 +42,33 @@ const handleCloseWithAction = ( e, specifiedAction ) => {
 
-const GuideImplementation = () => {
+const GuideImplementation = ( { isOpen, setCESPromptOpen } ) => {
+	const handleGotItClick = useCallback(
+		( e ) => {
+			handleCloseWithAction( e );
+			setCESPromptOpen( true );
+		},
+		[ setCESPromptOpen ]
+	);
```
❓ / 💅 

What i the rationale behind using `handleGotItClick` separated. Instead of adding 

```
if ( action === CTA_CREATE_ANOTHER_CAMPAIGN ) {
		getHistory().push( getCreateCampaignUrl() );
}

if ( action === CTA_CONFIRM) {
		setCESPromptOpen( true );
}
```
I guess is something related to the UseCallback? Not sure.

ianlin replied:
I think originally my thought was: the `handleCloseWithAction` is defined outside of the component `GuideImplementation`, we will need to pass the function prop `setCESPromptOpen()` to it and I think it could be inflexible if we have more custom actions. But I agree with you that using `handleCloseWithAction` to handle multiple actions would be better than using a separated `handleGotItClick`. What do you think if we move `handleCloseWithAction` inside the component so it can use the function prop `setCESPromptOpen()` directly?

puntope replied:
Ohhh, understood, didn't show that one. Thanks for the explanation.

>  What do you think if we move handleCloseWithAction inside the component 

Would be interesting to know why it is outside, wondering if there is an important reason for it, maybe @eason9487 knows. 

I would go ahead with your current solution and the maybe refactor after knowing that. It's 100% a non blocking issue.  Since I don't want to block the PR I rely on you for the final decision regarding this. Again, looks good now as it's.

eason9487 replied:
Those click handlers were not dependent on props and states,  so to avoid their references changing all the time, I put them outside then. It's ok to move them inside the component.

💅 
BTW, I'd suggest adding a callback prop, such as `onGuideRequestClose( action )`, to this component and moving the process of `handleCloseWithAction` handler to the upper component. Therefore, it could decouple the `<CustomerEffortScorePrompt>` from this guide, and the logic of the opening/closing this guide could also be centralized on the outside. In this way, as with product feed index, the similar logic is implemented at dashboard index.

ianlin replied:
@eason9487 Great suggestion. Added in ~bc3aaef~ 739e897.

puntope commented on js/src/components/customer-effort-score-prompt/index.js:
```diff
@@ -0,0 +1,92 @@
+	);
+};
+
+CustomerEffortScorePrompt.propTypes = {
+	/**
+	 * The text to be displayed on the CES prompt
+	 */
+	label: PropTypes.string.isRequired,
+	/**
+	 * Context property to be sent when calling recordEvent
+	 */
+	eventContext: PropTypes.string.isRequired,
+};
```
💅 

Asking around about PropTypes since I never saw this in any of our projects, would be lovely to have some kind of discussion before implementing it. 

p1639742969012200-slack-CK365S85V

ianlin replied:
Gotcha, I should have checked whether we use PropTypes in the codebase before implementing it. Thanks for bringing this up and the slack discussion was amazing! @tomalec 👍

Updated in 006b2ac.

puntope commented:
> Prompt 1, Flow 1  ✅ 

Going to the page `/wp-admin/admin.php?page=wc-admin&path=/google/product-feed&guide=submission-success` 

Behaviour and visual testing these for now and looks good ✅  

- When modal is visible, then CES isn't ✅ 
- Dismiss the modal with X (CES shows and records event) ✅ 
- Dismiss the modal with Cancel button (CES shows and records event) ✅ 
- Dismiss the modal with MaybeLater button (CES shows and records event) ✅
- Disable "Allow tracking" in Settings (CES not showing up) ✅ 
- Reproduced video expected behaviour ✅

puntope commented:
> Prompt 1 Flow 2 

🚧  

I could not complete the test instructions successfully (I take as granted that `Follow the first three steps above` means `Follow the first two steps above` because I'm not able to click the new tab with the modal dismissed) 

Followed steps:

1. Go to the product feed page with the success modal opened: /wp-admin/admin.php?page=wc-admin&path=/google/product-feed&guide=submission-success
2. Enable event debugging by typing localStorage.setItem('debug', 'wc-admin*') in your web's dev console
3. Go to the product feed page in a new tab by click Manage and edit your product feed in WooCommerce
4. I see the CES prompt and the recordEvent ✅ 
5.  Go back to the original tab and dismiss the success modal, the CES is there ❌  


https://user-images.githubusercontent.com/5908855/146567062-a3dbb974-a3d8-42fb-ad92-25df38014d5e.mov

puntope commented:
> Prompt 2 ✅ 

1. Went to `/wp-admin/admin.php?page=wc-admin&path=/google/dashboard&guide=campaign-creation-success`
2. Clicked on "Got it"
3. CES was shown, event was tracked ✅ 
4. Repeated 1
5. Dismissed modal
6. CES was not shown ✅ 
7. With Allow tracking unchecked CES was not shown ✅

puntope commented:
> CES Events 🚧  /❓ 

- When I click on feedback, it records `wcadmin_gla_ces_modal_open` ✅ 
- When I send feedback, it records `wcadmin_gla_ces_feedback`with the right score and empty comments.✅ 
- When I close CES, it records `wcadmin_gla_ces_snackbar_closed` ✅ 
- When CES shows, it records `wcadmin_gla_ces_snackbar_open` ✅ 


🚧  /❓ 

In product feed, context was `gla-setup` ✅  but parent modal and other tracks shows `submission-success` as context in the same scope.

In campaigns setup, context was `create-ad-campaign` but the parent modal context was `campaign-creation-success` 

I guess we should maintain consistency between them. Since it's exactly the same context. Thoughts?

puntope commented on js/src/product-feed/index.js:
```diff
@@ -10,18 +11,45 @@ import NavigationClassic from '.~/components/navigation-classic';
+	// Show submission success guide modal by visiting the path with a specific query `guide=submission-success`.
+	// For example: `/wp-admin/admin.php?page=wc-admin&path=%2Fgoogle%2Fproduct-feed&guide=submission-success`.
+	const isSubmissionSuccessOpen =
+		getQuery().guide === GUIDE_NAMES.SUBMISSION_SUCCESS;
```
💅  Again, Personal opinion 100% not blocking

I saw that `getQuery()` return always Object somehow.  But maybe `getQuery()?.guide` would be safer, since it's an external dependency.

ianlin replied:
Great suggestion! Added in 19ff496.

ianlin committed:
Add docs for event records of CES prompts

ianlin committed:
Remove using PropTypes

Ref: p1639742969012200-slack-CK365S85V

ianlin committed:
Add optional chaining for getQuery()

ianlin committed:
Update event context for CES prompts to match their parents' context

ianlin commented:
@puntope 

> I guess we should maintain consistency between them. Since it's exactly the same context. Thoughts?

Good catch! I have no idea why I created a different context.. thanks! Added in 3c9b2f6

findingsimple commented:
Nice work @ianlin ❤️

ianlin committed:
Add a component state canCESPromptOpen in ProductFeed

Reason to change:

There was a bug

- Success modal is displayed, CES prompt is not shown
- Open a new tab to the product feed page, CES prompt is shown
- Go back to the original tab and dismiss the success modal, the CES
	prmopt is shown, which is what we don't want.

To solve it, moving `removeCESPromptFlagFromLocal` from
`onNoticeDismissed` and `onModalShown` to only `onNoticeShown` is not
enough, as it leads to another problem:

- Success modal is displayed, CES prompt is not shown
- Open a new tab to the product feed page, CES prompt is shown
- Click `Give feedback` but the CES modal is not displayed

This is because we change to remove the CES flag from local storage when
the CES prompt is shown. When it happens the ProductFeed will re-render
and the CES prmopt will not be rendered because the flag was removed.
Then clicking `Give feedback` won't render the modal.

To solve this we add a component state `canCESPromptOpen` in ProductFeed
component, rather than check the local storage flag directly. Then we
add the logic to update the `canCESPromptOpen` state only when its value
is false.

ianlin commented:
@puntope 

> 5. Go back to the original tab and dismiss the success modal, the CES is there ❌

Good catch as well! I have updated this in 07d85e4.

To explain this:

Originally the `removeCESPromptFlagFromLocal` function is called when the CES modal is shown or is dismissed. Moving `removeCESPromptFlagFromLocal` from `onNoticeDismissed`, `onModalShown` to only `onNoticeShown` solves the problem, but it leads to another problem:

- Success modal is displayed, CES prompt is not shown
- Open a new tab to the product feed page, CES prompt is shown
- Still in the new tab, clicking `Give feedback` but the CES modal is not displayed

This is because we change to remove the CES flag from local storage when the CES prompt is shown. When it happens the ProductFeed will re-render and the CES prmopt will not be rendered because the flag was removed. Then clicking `Give feedback` won't render the modal.

To solve this we add a component state `canCESPromptOpen` in `ProductFeed` component, rather than check the local storage flag directly. Then we add the logic to update the `canCESPromptOpen` state only when its value is false, and it solves the problem and everything works as expected now.

Here's the video to demonstrate the fixed behaviour:

https://user-images.githubusercontent.com/914406/146749980-ecbd7dc2-e637-4344-bf90-b770aeb9bb9c.mov

puntope commented on js/src/components/customer-effort-score-prompt/index.js:
```diff
@@ -0,0 +1,106 @@
+ * @param {string} props.label Text to be displayed in the CES notice and modal.
+ * @return {JSX.Element} Rendered element.
+ */
+const CustomerEffortScorePrompt = ( { eventContext, label } ) => {
```
💅  Non blocking

I guess we can add here

`if ( ! isWCTracksEnabled() ) return null;`  Since it is a common check for all the prompts, right?

And then remove them in `js/src/dashboard/campaign-creation-success-guide/index.js` and `js/src/product-feed/index.js`

puntope commented on js/src/components/customer-effort-score-prompt/index.js:
```diff
@@ -0,0 +1,106 @@
+/**
+ * CES feedback recorded
+ *
+ * @event gla_ces_feedback
```
```suggestion
 * @event gla_ces_feedback
 * @property {number} score The provided feedback score.
 * @property {string} comments If a comment is provided.
```

puntope replied:
💅

puntope commented on js/src/dashboard/campaign-creation-success-guide/index.js:
```diff
@@ -115,10 +138,26 @@ const GuideImplementation = () => {
+				isOpen={ isOpen }
+				setCESPromptOpen={ setCESPromptOpen }
+			/>
+			{ isCESPromptOpen && wcTracksEnabled && (
```
💅 

See my previous comment regarding `isWCTracksEnabled();`

puntope commented on js/src/product-feed/index.js:
```diff
@@ -10,18 +12,54 @@ import NavigationClassic from '.~/components/navigation-classic';
+	// For example: `/wp-admin/admin.php?page=wc-admin&path=%2Fgoogle%2Fproduct-feed&guide=submission-success`.
+	const isSubmissionSuccessOpen =
+		getQuery()?.guide === GUIDE_NAMES.SUBMISSION_SUCCESS;
+
+	const wcTracksEnabled = isWCTracksEnabled();
```
💅 

See comment regarding  `isWCTracksEnabled();`

puntope commented on js/src/utils/localStorage.js:
```diff
@@ -0,0 +1,22 @@
+
+const localStorage = {
+	get( key ) {
+		return localStorageExists ? window.localStorage.getItem( key ) : null;
```
💅  What do you think about...

`return window?.localStorage?.getItem( key )`  or `return window?.localStorage?.getItem( key ) ?? null` (in case we need null as return for a reason). 

That way we can remove both conditionals and the localStorageExists computed function

puntope commented on js/src/product-feed/index.test.js:
```diff
@@ -0,0 +1,180 @@
+			} );
+		} );
+
+		describe( 'And both canCESPromptOpenLocal and wcTracksEnabled are true', () => {
```
💅 

RE-  comments regarding `isWCTracksEnabled` 

It will also simplify the test. We can have a separated test for the CESPrompt checking that and here only test LocalStorage / SubmissionSuccessGuide

puntope reviewed (approved):
✅   LGTM. Good work!

All the blocking issues were addressed and this PR looks good for me

Prompt 1 test ✅ 
Prompt 1.2 test ✅ 
Prompt 2 thank test ✅ 
Unit tests ✅   

💅  Just a few pending comments out there

ianlin committed:
Decouple CustomerEffortScorePrompt and the CampaignCreationSuccessGuide

ianlin committed:
Add @woocommerce/settings dependency (DEWP) in Jest config

Ref: https://github.com/woocommerce/woocommerce-admin/issues/7810

ianlin committed:
Unit test on Dashboard rather than CampaignCreationSuccessGuide
